src/main/java/com/sunnylabs/tracegenerator/Configuration.java

```diff
@@ -0,0 +1,282 @@
+package com.sunnylabs.tracegenerator;
+
+import com.wavefront.java_sdk.com.google.common.base.Strings;
+import lombok.Data;
+import org.springframework.lang.NonNull;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.constructor.Constructor;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.*;
+
+public class Configuration {
+    private final RawConfig raw;
+    private int servicesPerApp = 50;
+    private int internalCallsPerApp = 3;
+
+    public Configuration(InputStream stream) {
+        Yaml yaml = new Yaml(new Constructor(RawConfig.class));
+        raw = getRawConfig(stream, yaml);
+        setDefaults();
+        checkCallGraph();
+    }
+
+    private void checkCallGraph() {
+        for (Application app : raw.applications.values()) {
+            if (app.getServices() == null || app.getServices().isEmpty()) {
+                continue;
+            }
+            for (Service svc : app.getServices().values()) {
+                List<Operation> ops = svc.getOperations();
+                if (ops == null || ops.isEmpty()) {
+                    continue;
+                }
+                for (Operation op : ops) {
+                    checkCalls(op, Collections.emptyList());
+                }
+            }
+        }
+    }
+
+    private void checkCalls(Operation operation, List<Operation> callGraph) {
+        List<Operation> graph = new ArrayList<>(callGraph);
+        graph.add(operation);
+        for (Operation o : operation.getCalls()) {
+            if (graph.contains(o)) {
+                throw new IllegalArgumentException(String.format("Operation %s.%s.%s has circular" +
```

Comment:
This helper method only gets used in the constructor, which is only used at startup. The wordlist-based generation tries to avoid circular references, so this is mostly to validate user-provided config.